### DISCUSSION TITLE
Can you create a z.enum from an existing string literal union?

### DISCUSSION THREAD

**Discussion by Je12emy** [colinhacks/zod#1920] [Q&A] [answered]
Take the following Prisma Schema.

```prisma
model Patient {
    id        String   @id @default(cuid())
    firstName String   
    lastName  String
    bloodType BoodType
}

enum BoodType {
    A
    B
    AB
    O
}
```
Since I'm using tRPC, I would like to infer some query's output type, which returns this Patient type.

```ts
export type Patient = NonNullable<RouterOutput["patient"]["byId"]>
export type BloodType = Pick<Patient, "bloodType">["bloodType"]
//                     ^? type BloodType = "A" | "B" | "AB" | "O"
```

So I would like to create a `z.enum` based on this `BlodType` string literal, is this possible?

**Comment by JacobWeisenburger** [comment, marked as answer]
> ```ts
> export type Patient = NonNullable<RouterOutput["patient"]["byId"]>
> export type BloodType = Pick<Patient, "bloodType">["bloodType"]
> //                     ^? type BloodType = "A" | "B" | "AB" | "O"
> ```

I'm not very familiar with Prisma, so please forgive my ignorance. It looks like you are only showing types. But `z.enum` requires a runtime value. Is there a way to get the values of the `BloodType` enum at runtime to pass into `z.enum`?

**Reply by Je12emy** [reply]
Sorry, I forgot to mention that these types are extracted from a tRPC router, not from prisma itself. Here are the related docs: https://trpc.io/docs/infer-types

As for my solution, I ended up duplicating this array for until I figure something out.

**Reply by JacobWeisenburger** [reply]
`z.enum` requires a runtime value to be passed in. So if all you have are types from tRPC or Prisma, then I don't think you can make an `z.enum` from that. Is there a way to get the values of the `BloodType` enum at runtime?

**Comment by robinelvin** [comment]
I have a similar use case. I need to validate against a literal union that is in a library that I don't control (dayjs)

```typescript
type plugin.DurationUnitType = "s" | "month" | "week" | "millisecond" | "second" | "minute" | "hour" | "day" | "year" | "milliseconds" | "seconds" | "minutes" | "hours" | "days" | "months" | "years" | "d" | "D" | ...```

Is there a way to validate a field against this union?

**Reply by JacobWeisenburger** [reply]
zod requires a runtime value to be passed into a schema. So if all you have are types, then I don't think you can make a zod schema from that. Is there a way to get the values of the `plugin.DurationUnitType` union at runtime?

Perhaps something like this could help?
https://github.com/fabien0102/ts-to-zod

**Reply by robinelvin** [reply]
I think I could have used ts-to-zod but I'm reluctant to add another library and have another build step to generate code.
In the end I followed the pattern described [here](https://dev.to/hansott/how-to-check-if-string-is-member-of-union-type-1j4m) and used the type guard function in a `z.custom` function. It's a pity Typescript doesn't have a way to type check string unions at runtime.

```typescript
const Units = [
    'milliseconds',
    'seconds',
    'minutes',
    'hours',
    'days',
    'weeks',
    'months',
    'years',
    // rest of literals here ...
] as const;
type UnitsTuple = typeof Units;
type Unit = UnitsTuple[number];

const PolicySchema = z.object({
    maxAge: z.coerce.number().positive(),
    maxAgeUnit: z.custom<Unit>((value: unknown): value is Unit =>
        Units.includes(value as Unit)
    ),
});
```
